### PR TITLE
Reduce portal DB IT concurrency (#2399)

### PR DESCRIPTION
**Pull request by noah-aviel-dove**
Author (check before every review)

- [x] PR title references issue
- [x] Title of main commit references issue
- [x] PR is linked to Zenhub issue
- [x] Added `DCP/1` label                                   <sub>or this PR does not target an `hca/*` branch</sub>
- [x] Created issue to track porting these changes          <sub>or this PR does not need porting</sub> 
- [x] Added `r` tag to commit title                         <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                           <sub>or this PR does not require reindexing</sub>
- [x] Freebies are blocked on this PR                       <sub>or there are no freebies in this PR</sub>
- [x] Freebies are referenced in commit titles              <sub>or there are no freebies in this PR</sub>
- [x] This PR is blocked by previous PR in chain            <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR                <sub>or this PR is not chained to another PR</sub>
- [x] Ran `make requirements_update`                        <sub>or this PR leaves requirements*.txt, common.mk and Makefile untouched</sub>
- [x] Added `R` tag to commit title                         <sub>or this PR leaves requirements*.txt untouched</sub>
- [x] Added `reqs` label to PR                              <sub>or this PR leaves requirements*.txt untouched</sub>
- [x] Documented upgrading of deployments in UPGRADING.rst  <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                         <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                           <sub>or this PR does not require upgrading</sub>
- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased branch on `develop`

Primary reviewer (before pushing merge commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Build passes in `sandbox`                             <sub>or added `no sandbox` label</sub>
- [x] Reindexed `sandbox`                                   <sub>or this PR does not require reindexing</sub>
- [x] Added PR reference to merge commit
- [x] Moved linked issue to Merged
- [ ] Pushed merge commit to Github

Primary reviewer (after pushing merge commit)

- [ ] Moved freebies to Merged column                       <sub>or there are no freebies in this PR</sub> 
- [ ] Shortened chain                                       <sub>or this PR is not the base of another PR</sub>
- [ ] Verified that `N reviews` labelling is accurate
- [ ] Commented on demo expectations                        <sub>or labeled as `no demo`</sub>
- [ ] Pushed merge commit to Gitlab                         <sub>or this changes can be pushed later, together with another PR</sub>
- [ ] Reindexed `dev`                                       <sub>or this PR does not require reindexing</sub>
- [ ] Deleted PR branch from Github and Gitlab
- [ ] Unassign reviewer from PR

### CODE REVIEW

**Review by achave11 — APPROVED** [review]
Looks good!

File: UPGRADING.rst [hunk 1]
```diff
@@ -10,6 +10,18 @@ branch that does not have the listed changes, the steps would need to be
...
+#2399 Reduce portal DB IT concurrency
+=====================================
+
+Delete the existing portal database files ::
+
+    s3://edu-ucsc-gi-singlecell-azul-config-dev.us-east-1/azul/{your-deployment}/portals/*.json
+
+and their corresponding entries in the DynamoDB object version table ::
+
+    azul-object_versions-{your-deployment}
```
Comment (hannes-ucsc) [inline UPGRADING.rst:22]: Are you expecting every team member to figure out how to do this on their own? Seems lazy.

File: UPGRADING.rst [hunk 2]
```diff
@@ -10,6 +10,18 @@ branch that does not have the listed changes, the steps would need to be
 reverted. This is all fairly informal and loosely defined. Hopefully we won't
 have too many entries in this file.
 
+#2399 Reduce portal DB IT concurrency
+=====================================
+
+Delete the existing portal database files ::
+
+    s3://edu-ucsc-gi-singlecell-azul-config-dev.us-east-1/azul/{your-deployment}/portals/*.json
```
Comment (hannes-ucsc) [inline UPGRADING.rst:18]: Same here

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
Like I said in the other PR, the chain needs to be inverted.

File: UPGRADING.rst (hunk 1 above)
Comment (noah-aviel-dove) [reply UPGRADING.rst]: I've elaborated on the instructions.

File: UPGRADING.rst [hunk 3]
```diff
@@ -10,6 +10,18 @@ branch that does not have the listed changes, the steps would need to be
...
+
+and their corresponding entries in the DynamoDB object version table ::
+
+    azul-object_versions-{your-deployment}
+
+To ensure that no pollution persists from previous IT failures due to #2400.
```
Comment (hannes-ucsc) [inline UPGRADING.rst]: Fragment.

File: UPGRADING.rst [hunk 4]
```diff
@@ -10,6 +10,40 @@ branch that does not have the listed changes, the steps would need to be
 reverted. This is all fairly informal and loosely defined. Hopefully we won't
 have too many entries in this file.
 
+#2399 Reduce portal DB IT concurrency
+=====================================
+
+Delete the existing portal databases and their corresponding version tracking
```
Comment (hannes-ucsc) [inline UPGRADING.rst:16]: I think there should be a script for this. It's too involved.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: UPGRADING.rst (hunk 3 above)
Comment (noah-aviel-dove) [reply UPGRADING.rst]: This comment appears to be in reference to an older commit. The fragment in question has already been `fixup`-ed.

File: UPGRADING.rst (hunk 4 above)
Comment (noah-aviel-dove) [reply UPGRADING.rst]: The script overwrites the database with its default value (hardcoded in the plugin) rather than deleting it, since this requires fewer changes to the `PortalService` class.

File: src/azul/portal_service.py [hunk 5]
```diff
@@ -41,6 +43,20 @@ def __init__(self):
...
+        return config.portal_db_bucket
+
+    @property
+    def object_key(self):
+        return config.portal_db_object_key(self.data_source)
+
+    @cached_property
+    def data_source(self):
+        # FIXME https://github.com/DataBiosphere/azul/issues/2716
+        catalog = config.default_catalog
```
Comment (achave11) [inline src/azul/portal_service.py:57]: I think this variable could be inlined, up to you. If the way it is now is clearer, disregard my comment.

**Review by achave11 — APPROVED** [review]
LGTM, approved!

File: src/azul/portal_service.py (hunk 5 above)
Comment (noah-aviel-dove) [reply src/azul/portal_service.py:57]: That's the idea, although mostly I was following Hannes's precedent from https://github.com/DataBiosphere/azul/blob/dc202ee2c273cee1bff823ee2b930d5d8fe0addc/src/azul/portal_service.py#L155

File: src/azul/portal_service.py [hunk 6]
```diff
@@ -41,6 +43,20 @@ def __init__(self):
...
+    def bucket(self):
+        return config.portal_db_bucket
+
+    @property
+    def object_key(self):
+        return config.portal_db_object_key(self.data_source)
+
+    @cached_property
+    def data_source(self):
+        # FIXME https://github.com/DataBiosphere/azul/issues/2716
```
Comment (hannes-ucsc) [inline src/azul/portal_service.py:56]: Fix formatting and add one-liner comment, potentially just the title of the issue.

File: src/azul/portal_service.py [hunk 7]
```diff
@@ -70,6 +86,13 @@ def callback(portal_db):
         self._crud(callback)
         return result
 
+    @cached_property
+    def default_db(self) -> JSONs:
+        # FIXME https://github.com/DataBiosphere/azul/issues/2716
```
Comment (hannes-ucsc) [inline src/azul/portal_service.py:91]: Same here

File: src/azul/portal_service.py [hunk 8]
```diff
@@ -41,6 +43,20 @@ def __init__(self):
...
+    @property
+    def bucket(self):
+        return config.portal_db_bucket
+
+    @property
+    def object_key(self):
+        return config.portal_db_object_key(self.data_source)
+
+    @cached_property
+    def data_source(self):
```
Comment (hannes-ucsc) [inline src/azul/portal_service.py:55]:
```suggestion
    def catalog_source(self):
```

here and in comments, docstrings etc

File: src/azul/portal_service.py [hunk 9]
```diff
@@ -42,6 +43,20 @@ def __init__(self):
         self.client = aws.client('s3')
         self.version_service = VersionService()
 
+    @property
+    def bucket(self):
```
Comment (hannes-ucsc) [inline src/azul/portal_service.py:47]: This should be in a separate commit.

**Review by hannes-ucsc — COMMENTED** [review]
Don't bother with fixups on the next review cycle. And be sure to structure the history cleanly.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
Clicked the wrong button. Previous review was meant to request changes instead of just commenting.

File: test/service/test_portal_service.py [hunk 10]
```diff
@@ -62,20 +62,21 @@ def plugin_db(self) -> JSONs:
...
-        {
+    @property
+    def multiplex_db(self) -> JSONs:
+        return [{
             "integrations": [
                 # this should be flattened
                 {
                     "entity_ids": {
-                        config.dss_deployment_stage: ["good"],
+                        self.portal_service.catalog_source: ["good"],
```
Comment (hannes-ucsc) [inline test/service/test_portal_service.py:72]: This change is in the wrong commit. It uses a property that's defined in a later commit.

File: test/service/test_portal_service.py [hunk 11]
```diff
@@ -62,20 +62,21 @@ def plugin_db(self) -> JSONs:
...
                         "other": ["bad"],
                     }
                 },
                 # this should be removed (entity_ids defined but missing for current stage)
                 {
                     "entity_ids": {
-                        config.dss_deployment_stage: [],
+                        self.portal_service.catalog_source: [],
```
Comment (hannes-ucsc) [inline test/service/test_portal_service.py:79]: This change is in the wrong commit. It uses a property that's defined in a later commit.

File: test/service/test_portal_service.py [hunk 12]
```diff
@@ -84,48 +85,47 @@ def plugin_db(self) -> JSONs:
...
                 }
             ]
-        }
-    ]
+        }]
 
-    demultiplex_db = [
-        {
+    @property
+    def demultiplex_db(self) -> JSONs:
```
Comment (hannes-ucsc) [inline test/service/test_portal_service.py:91]: Why was this converted from an attribute to a property? If for consistency, then please revert. This is less efficient and more verbose that the current implementation. If to enable mutability of the returned value, the redeclare as MutableJSON. If for other reasons, please explain those here.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

**Review by hannes-ucsc — APPROVED** [review]

**Comment by hannes-ucsc** [comment]
For demo, show integration test output on sandbox. Show that it's skipped on `dev`. Show the contents of the registration database on sandbox before and after the test.

**Comment by hannes-ucsc** [comment]
@amarjandu note the `chain` and `upgrade` labels. Both `sandbox` and `dev` will need to be upgraded.
